### ISSUE TITLE
Generic tuple with a member's type depending on another member

### ISSUE THREAD

**Issue by MKRhere** [microsoft/TypeScript#37349]
**TypeScript Version:**  3.9.0-dev.20200310

**Search Terms:**

Tuple with a member's type depending on another member

**Code**

```ts
type Predicate = (x: any) => boolean;
const isString = (x: any): x is string => typeof x === "string";

type GuardedType<T> = T extends (x: any) => x is infer T ? T : never;

/* Case 1 */
const f = <T extends Predicate, U extends GuardedType<T>>(x: T, y: U): any => x;

/* Case 2 */
const g = <T extends Predicate, U extends [GuardedType<T>]>(x: T, y: U): any => x;

/* Case 3 */
const h = <T extends Predicate, U extends [T, GuardedType<T>]>(x: U): any => x;

f(isString, ""); // OK
g(isString, [""]); // OK
h([isString, ""]); // ERROR: Type 'string' is not assignable to type 'never'. ts(2322)
```

**Expected behavior:**

All 3 should work

**Actual behavior:**

Third scenario seems to resolve `GuardedType<T>` before `T` is known from call-site, so it uses the generic constraint (`extends Predicate`) and ends up inferring `GuardedType<T>` as `never`.

**Playground Link:** [TypeScript Playground](https://www.typescriptlang.org/play/?ts=3.9.0-dev.20200310#code/C4TwDgpgBACgThAJgSwMYENjQLxQBQAeAXFOgHYgCUU2AfFAEYD2TANhOQNwCwAUKkzIBnYFGRCAysDjIyAcxr5ipCpRIExQqCJnya9UJCYAzKBuwWoAIh2y5Vnrz6HoAcQCu6OIiQAVcBAAPL70uL5QEARYZIhahCTkVPpmmmJkxhBwUOEA-NlQJGQQAG6ZjgD0AFRQAMLoQtAAjFCV5XwCwqKmuMERURAxWvBIaJgQADRQAKp90bFQHl4+iP6QwbS08dmTICRTaiogyQQV1XUNUABMLW38giJQCj3hkXNDCCgYWJMzrwPzAG1Ft4-AF1gBdTbKXw7PYHRLHU61erQADMN3a91EAAtFL0-oNYB9Rt9prN-loATCFp4QSswSFIVt9gkKIi+HxjHhxFJdHJJlYrJROFByuUoAB5ADSfDk3Mk0jskwBgvBwtF4ulfGxeABPMV8gFVjVIrFUAAogAlS0Sy18IA)

**Related issues:**

https://stackoverflow.com/questions/57100330/typescript-type-of-tuple-member-dependent-on-other

The solution here works similar to:

```ts
const f = <T extends Predicate, U extends GuardedType<T>>([ x: T, y: U ]): any => x;
```

This already works fine, but in my actual use-case, the tuple is further used to infer types in subsequent generic constraints, so it has to be part of the generic constraint.

I spent a lot of time messing with this, so let me know if I'm the one missing something, thanks :)

**Comment by RyanCavanaugh** [comment]
I can't tell how you want these functions to behave since there are no examples. Did you want something like this?
```ts
const i = <T>(arg: [(arg: any) => arg is T, T]) => arg;
// OK
i([isString, ""]); 
```

**Comment by MKRhere** [comment]
Thanks for that; but it doesn't seem to work in my case. I tried to extract the bare parts, but I'll give you the full use-case to not XY.

[**Playground**](https://www.typescriptlang.org/play/?ts=3.9.0-dev.20200310&ssl=1&ssc=1&pln=18&pc=20#code/C4TwDgpgBACgThAJgSwMYENjQLxQBQAeAXFOgHYgCUU2AfFAEYD2TANhOQNwCwAUKJCgBxAK7o4iJABVwEADxT6uKVAgEsZRAGd8xUhWp0oBKMh3IyAMwhwoKgPx2oJMhABuNnv1lQAsujBRcW0FJSgAbygAbQBpUzIoAGsIECZLOwBdEiCJaVkFWIz6AF8vAWgpcQBzCGAFVXUITR14JDRMCCiMgBooACVakTgyMLwAOgnqrRJ-QLEJLVDDegHgIbIynwBVMmQmMikmAEkyLDgtCFRgPbI5LdGtho1tfRA+AEhHPESSLeWoNxMZCID4udw2ahqZ46b4kCzWWxHf6A4EfRxHUFQVweOBePjlKBgNggGhQOQfVooDBYHRQpovSntLBdbofVbrJ70nTkECs3jvADyONYTHQiE5zWijOpEC0vUqcBqdWlHTl-UGw1oPQ+QpsIrFtMakt1cH1iBZFOJADERGQrjcJS8djdDiczhd7fs5CazVoomQRABbBg2DJRACMRT4tDwHwmYyYwtF2hIPuTWj4-xg1ttno2fD4qH2WmApi0AGVgHALFVSYQSDzKCQTGYoCXq2Ra0ZymljDRsLgAETtmuDvG8dilomsEAkack3DjSaK6avf5TcfzvBRKJmSsdqo9PtGAgZSheAD0F77WgAFkwRKxxbf0B5GBAmvEEQhxegdAADEdO3-eISw4cVe3-bEbH-PggA)

**Code:**

```ts
type Predicate = (x: any) => boolean;
type GuardedType<T> = T extends (x: any) => x is infer T ? T : never;
type MapGuards<T> = { [K in keyof T]: GuardedType<T[K]> };
type Target<T extends Predicate[], Return> = (...args: MapGuards<T>) => Return;
type UnionToIntersection<U> = (U extends any
	? (k: U) => void
	: never) extends (k: infer I) => void
	? I
	: never;

type poly = <
	Predicates extends Predicate[],
	Return extends any,
	Overload extends [Predicates, Target<Predicates, Return>],
	Overloads extends Overload[],
	PolyFunction extends UnionToIntersection<Overloads[number][1]>
>(
	...overloads: Overloads
) => PolyFunction;

const isString = (x: any): x is string => typeof x === "string";

// -- Usage:

poly([[isString], x => x]);
// x should have been inferred as `string` instead of `never`
```

Basically, I've created a polymorphism library for TS, which infers the input types from the type guards themselves (or it should, but doesn't atm). I appreciate all the work you've done, hopefully I'm not wasting time :)

**Comment by MKRhere** [comment]
Note: Attempting to type Predicate as `<T>(x: any) => x is T>` causes `isString` to no longer be a Predicate; with the error: `'string' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'`, and x becomes `unknown`
